# Part 1 - Data Cleaning

In [1]:
# Loading libraries:

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Reading data files:
# using encoding = "ISO-8859-1" to avoid pandas encoding error

rounds = pd.read_csv("rounds2.csv",encoding="ISO-8859-1")
companies = pd.read_table("companies.txt",encoding="ISO-8859-1")

In [3]:
# Viewing the rounds file:

rounds.head()

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [4]:
# Inspecting the structure of rounds():

rounds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   company_permalink        114949 non-null  object 
 1   funding_round_permalink  114949 non-null  object 
 2   funding_round_type       114949 non-null  object 
 3   funding_round_code       31140 non-null   object 
 4   funded_at                114949 non-null  object 
 5   raised_amount_usd        94959 non-null   float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB


In [5]:
rounds.shape

(114949, 6)

In [6]:
# Inspecting the structure companies():

companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [7]:
# converting all the permalinks to the lowercase:

companies["permalink"] = companies["permalink"].str.lower()
companies.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [8]:
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   permalink      66368 non-null  object
 1   name           66367 non-null  object
 2   homepage_url   61310 non-null  object
 3   category_list  63220 non-null  object
 4   status         66368 non-null  object
 5   country_code   59410 non-null  object
 6   state_code     57821 non-null  object
 7   region         58338 non-null  object
 8   city           58340 non-null  object
 9   founded_at     51147 non-null  object
dtypes: object(10)
memory usage: 5.1+ MB


In [9]:
companies.shape

(66368, 10)

Ideally, the permalink column in the companies dataframe should be the unique_key of the table, having 66368 unique company names (links, or permalinks). Also, these 66368 companies should be present in the rounds file.

Let's first confirm that these 66368 permalinks (which are the URL paths of companies' websites) are not repeating in the column, i.e. they are unique.



In [10]:
# Look at the unique values:

len(companies.permalink.unique())

66368

Thus, there are 66368 unique companies in the table and permalink is the unique primary key. Each row represents a unique company.

Let's now check whether all of these 66368 companies are present in the rounds file, and if some extra ones are present.

In [11]:
# look at unique company names in rounds df
# note that the column name in rounds file is different (company_permalink)

len(rounds.company_permalink.unique())

90247

There seem to be 90247 unique values of company_permalink, whereas we expected only 66368. May be this is because of uppercase/lowercase issues.

Let's convert the column to lowercase and look at unique values again

In [12]:
rounds["company_permalink"]=rounds["company_permalink"].str.lower()
rounds.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [13]:
len(rounds.company_permalink.unique())

66370

There seem to be 2 extra permalinks in the rounds file which are not present in the companies file. Let's hope that this is a data quality issue, since if this were genuine, we have two companies whose investment round details are available but their metadata (company name, sector etc.) is not available in the companies table.


Let's have a look at the company permalinks which are in the 'rounds' file but not in 'companies'.

In [14]:
# companies present in rounds file but not in (~) companies file
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  \
77                                /organization/10â°north   
729                   /organization/51wofang-无忧我房   
2670                           /organization/adslinked™   
3166    /organization/aesthetic-everythingâ®-social-ne...   
3291               /organization/affluent-attachã©-club-2   
...                                                   ...   
110545                    /organization/whodat’s-spaces   
113839                 /organization/zengame-ç¦
游科技   
114946                               /organization/ãeron   
114947                             /organization/ã”asys-2   
114948  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...   

                                funding_round_permalink   funding_round_type  \
77      /funding-round/b41ff7de932f8b6e5bbeed3966c0ed6a  equity_crowdfunding   
729     /funding-round/346b9180d276a74e0fbb2825e66c6f5b              venture   
2670    /funding-round/449ae54bb63c768c232955ca6911dee4                 seed   
3166    /funding-round/62593455f1a69857ed05d5734cc04132  equity_crowdfunding   
3291    /funding-round/626678bdf1654bc4df9b1b34647a4df1                 seed   
...                                                 ...                  ...   
110545  /funding-round/d5d6db3d1e6c54d71a63b3aa0c9278e6                 seed   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf                 seed   
114946  /funding-round/59f4dce44723b794f21ded3daed6e4fe              venture   
114947  /funding-round/35f09d0794651719b02bbfd859ba9ff5                 seed   
114948  /funding-round/af942869878d2cd788ef5189b435ebc4                grant   

       funding_round_code   funded_at  raised_amount_usd  
77                    NaN  12-08-2014                NaN  
729                     A  06-07-2015          5000000.0  
2670                  NaN  29-09-2014           100000.0  
3166                  NaN  12-10-2014                NaN  
3291                  NaN  15-10-2014           100000.0  
...                   ...         ...                ...  
110545                NaN  28-10-2014            30000.0  
113839                NaN  17-07-2010                NaN  
114946                  A  01-08-2014                NaN  
114947                NaN  01-01-2015            18192.0  
114948                NaN  01-10-2013            14851.0  

[74 rows x 6 columns]

All the permalinks have weird non-English characters. Let's see whether these characters are present in the original df as well

In [15]:
# Looking at the indices with wierd characters:

rounds_original = pd.read_csv("rounds2.csv",encoding="ISO-8859-1")
rounds_original.iloc[[29597, 31863, 45176, 58473], :]

company_permalink  \
29597                       /ORGANIZATION/E-CÊBICA   
31863  /ORGANIZATION/ENERGYSTONE-GAMES-ǁµÇŸ³Æ¸¸Æˆ   
45176          /organization/huizuche-com-惠ç§ÿ车   
58473        /ORGANIZATION/MAGNET-TECH-Ç£ÇŸ³Ç§‘ÆŠ€   

                               funding_round_permalink funding_round_type  \
29597  /funding-round/8491f74869e4fe8ba9c378394f8fbdea               seed   
31863  /funding-round/b89553f3d2279c5683ae93f45a21cfe0               seed   
45176  /funding-round/8f8a32dbeeb0f831a78702f83af78a36               seed   
58473  /funding-round/8fc91fbb32bc95e97f151dd0cb4166bf               seed   

      funding_round_code   funded_at  raised_amount_usd  
29597                NaN  01-02-2015                NaN  
31863                NaN  09-08-2014                NaN  
45176                NaN  18-09-2014                NaN  
58473                NaN  16-08-2014          1625585.0

The company weird characters appear when you import the data file. To confirm whether these characters are actually present in the given data or whether python has introduced them while importing into pandas, let's have a look at the original CSV file in Excel.

Thus, this is most likely a data quality issue we have introduced while reading the data file into python. Specifically, this is most likely caused because of encoding.

First, let's try to figure out the encoding type of this file. Then we can try specifying the encoding type at the time of reading the file. The chardet library shows the encoding type of a file.

In [16]:
import chardet

rawdata = open("rounds2.csv","rb").read()
result = chardet.detect(rawdata)
char_encode = result["encoding"]
print(char_encode)

Windows-1254


In [17]:
print(result)

{'encoding': 'Windows-1254', 'confidence': 0.4186155476629225, 'language': 'Turkish'}


In [18]:
rounds["company_permalink"] = rounds.company_permalink.str.encode("utf-8").str.decode("ascii","ignore")
rounds.loc[~rounds["company_permalink"].isin(companies["permalink"]),:]

company_permalink  \
77                                  /organization/10north   
729                               /organization/51wofang-   
2670                              /organization/adslinked   
3166    /organization/aesthetic-everything-social-network   
3291                 /organization/affluent-attach-club-2   
...                                                   ...   
110545                       /organization/whodats-spaces   
113839                             /organization/zengame-   
114946                                 /organization/eron   
114947                               /organization/asys-2   
114948  /organization/novatiff-reklam-ve-tantm-hizmetl...   

                                funding_round_permalink   funding_round_type  \
77      /funding-round/b41ff7de932f8b6e5bbeed3966c0ed6a  equity_crowdfunding   
729     /funding-round/346b9180d276a74e0fbb2825e66c6f5b              venture   
2670    /funding-round/449ae54bb63c768c232955ca6911dee4                 seed   
3166    /funding-round/62593455f1a69857ed05d5734cc04132  equity_crowdfunding   
3291    /funding-round/626678bdf1654bc4df9b1b34647a4df1                 seed   
...                                                 ...                  ...   
110545  /funding-round/d5d6db3d1e6c54d71a63b3aa0c9278e6                 seed   
113839  /funding-round/6ba28fb4f3eadf5a9c6c81bc5dde6cdf                 seed   
114946  /funding-round/59f4dce44723b794f21ded3daed6e4fe              venture   
114947  /funding-round/35f09d0794651719b02bbfd859ba9ff5                 seed   
114948  /funding-round/af942869878d2cd788ef5189b435ebc4                grant   

       funding_round_code   funded_at  raised_amount_usd  
77                    NaN  12-08-2014                NaN  
729                     A  06-07-2015          5000000.0  
2670                  NaN  29-09-2014           100000.0  
3166                  NaN  12-10-2014                NaN  
3291                  NaN  15-10-2014           100000.0  
...                   ...         ...                ...  
110545                NaN  28-10-2014            30000.0  
113839                NaN  17-07-2010                NaN  
114946                  A  01-08-2014                NaN  
114947                NaN  01-01-2015            18192.0  
114948                NaN  01-10-2013            14851.0  

[74 rows x 6 columns]

Now the permalink column looks clean. Let us look at the unique values in the rounds file

In [19]:
# Checking unique values for rounds

len(rounds["company_permalink"].unique())

66368

Observation:        
1. Initially we had 66370 unique values for rounds dataframe and now we are having 66368 for rounds. 
2. Now we have same number of unique values in companies and rounds dataframe.

We have looked at the rounds dataframe and resolved the encoding issue. What if there is a same encoding issue in companies 
dataframe. So lets look at those companies which are present in the companies file but not in the rounds file.

In [20]:
# companies that are present in the companies file but not in the rounds file:

companies.loc[~companies["permalink"].isin(rounds["company_permalink"]),:]

permalink  \
43                             /organization/10ã‚â°north   
426     /organization/51wofang-ã†â€” ã
â¿â§ã†ë†â€˜ã†ë†â¿   
1506                      /organization/adslinkedã¢â€žâ¢   
1775   /organization/aesthetic-everythingã¢â®-social-...   
1834            /organization/affluent-attachã£â©-club-2   
...                                                  ...   
63833               /organization/whodatã¢â‚¬â„¢s-spaces   
65778  /organization/zengame-ã‡â¦â€¦ã†â¸â¸ã‡â§â€˜ã†å¡â‚¬   
66365                             /organization/ãƒâeron   
66366                          /organization/ãƒâ€asys-2   
66367  /organization/ã„â°novatiff-reklam-ve-tanã¤â±tã...   

                                                    name  \
43                                           10°North   
426                   51wofang æ— å¿§æˆ‘æˆ¿   
1506                                    AdsLinkedâ„¢   
1775             Aesthetic Everything® Social Network   
1834                            Affluent Attaché Club   
...                                                  ...   
63833                             Whodat’s Spaces   
65778                ZenGame 禅游科技   
66365                                           ÁERON   
66366                                          Ôasys   
66367  İnovatiff Reklam ve Tanıtım Hizmetler...   

                          homepage_url  \
43                                 NaN   
426            http://www.51wofang.com   
1506          http://www.adslinked.com   
1775   http://aestheticeverything.com/   
1834   http://www.affluentattache.com/   
...                                ...   
63833                              NaN   
65778          http://www.zen-game.com   
66365             http://www.aeron.hu/   
66366             http://www.oasys.io/   
66367             http://inovatiff.com   

                                           category_list     status  \
43                                               Fashion  operating   
426                                                  NaN     closed   
1506                                Advertising|Internet  operating   
1775                                    Public Relations  operating   
1834                                         Hospitality  operating   
...                                                  ...        ...   
63833                                               Apps  operating   
65778                Internet|Mobile Games|Online Gaming     closed   
66365                                                NaN  operating   
66366  Consumer Electronics|Internet of Things|Teleco...  operating   
66367                 Consumer Goods|E-Commerce|Internet  operating   

      country_code state_code       region            city  founded_at  
43             CAN         ON      Toronto     Mississauga  05-01-2013  
426            NaN        NaN          NaN             NaN  01-01-2014  
1506           NaN        NaN          NaN             NaN         NaN  
1775           USA         CA  Los Angeles  Marina Del Rey  15-06-2009  
1834           USA         CA  Los Angeles   Beverly Hills         NaN  
...            ...        ...          ...             ...         ...  
63833          NaN        NaN          NaN             NaN         NaN  
65778          NaN        NaN          NaN             NaN  17-07-2010  
66365          NaN        NaN          NaN             NaN  01-01-2011  
66366          USA         CA  SF Bay Area   San Francisco  01-01-2014  
66367          NaN        NaN          NaN             NaN         NaN  

[68 rows x 10 columns]

Obervation: Companies dataframe also has special characters and we need to clean that first.

In [25]:
# Removing the encoding from companies dataframe:

companies["permalink"] = companies['permalink'].str.encode("Windows-1254").str.decode("ascii","ignore")

In [29]:
# Rounds file:

rounds.to_csv("rounds_clean.csv",sep=",",index=False)

In [30]:
# Companies file:

companies.to_csv("companies_clean.csv",sep="\t",index=False)